pallets/automation-price/src/lib.rs

```diff
@@ -0,0 +1,700 @@
+
+//! # Automation time pallet
+//!
+//! DISCLAIMER: This pallet is still in it's early stages. At this point
+//! we only support scheduling two tasks per hour, and sending an on-chain
+//! with a custom message.
+//!
+//! This pallet allows a user to schedule tasks. Tasks can scheduled for any whole hour in the future.
+//! In order to run tasks this pallet consumes up to a certain amount of weight during `on_initialize`.
+//!
+//! The pallet supports the following tasks:
+//! * On-chain events with custom text
+//!
+
+#![cfg_attr(not(feature = "std"), no_std)]
+pub use pallet::*;
+
+// #[cfg(test)]
+// mod mock;
+
+// #[cfg(test)]
+// mod tests;
+
+mod benchmarking;
+pub mod weights;
+
+mod fees;
+pub use fees::*;
+
+use core::convert::TryInto;
+use cumulus_pallet_xcm::{Origin as CumulusOrigin};
+use frame_support::{
+	pallet_prelude::*, sp_runtime::traits::Hash, traits::Currency, transactional, BoundedVec,
+};
+use frame_system::{pallet_prelude::*, Config as SystemConfig};
+use pallet_timestamp::{self as timestamp};
+use scale_info::{TypeInfo, prelude::format};
+use sp_runtime::{
+	traits::{SaturatedConversion, Saturating},
+	Perbill,
+};
+use sp_std::{vec, vec::Vec};
+
+pub use weights::WeightInfo;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	pub type AccountOf<T> = <T as frame_system::Config>::AccountId;
+	pub type BalanceOf<T> =
+		<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+	type UnixTime = u64;
+
+	/// The struct that stores all information needed for a task.
+	#[derive(Debug, Eq, Encode, Decode, TypeInfo)]
+	#[scale_info(skip_type_params(T))]
+	pub struct Task<T: Config> {
+		owner_id: AccountOf<T>,
+		provided_id: Vec<u8>,
+		asset: Vec<u8>,
+		direction: u8,
+		trigger_percentage: u128,
+	}
+
+	/// Needed for assert_eq to compare Tasks in tests due to BoundedVec.
+	impl<T: Config> PartialEq for Task<T> {
+		fn eq(&self, other: &Self) -> bool {
+			self.owner_id == other.owner_id &&
+			self.provided_id == other.provided_id &&
+			self.asset == other.asset &&
+			self.direction == other.direction &&
+			self.trigger_percentage == other.trigger_percentage
+		}
+	}
+
+	impl<T: Config> Task<T> {
+		pub fn create_event_task(
+			owner_id: AccountOf<T>,
+			provided_id: Vec<u8>,
+			asset: Vec<u8>,
+			direction: u8,
+			trigger_percentage: u128,
+		) -> Task<T> {
+			Task::<T> { owner_id, provided_id, asset, direction, trigger_percentage }
+		}
+	}
+
+	#[derive(Debug, Encode, Decode, TypeInfo)]
+	#[scale_info(skip_type_params(T))]
+	pub struct TaskHashInput<T: Config> {
+		owner_id: AccountOf<T>,
+		provided_id: Vec<u8>,
+	}
+
+	impl<T: Config> TaskHashInput<T> {
+		pub fn create_hash_input(owner_id: AccountOf<T>, provided_id: Vec<u8>) -> TaskHashInput<T> {
+			TaskHashInput::<T> { owner_id, provided_id }
+		}
+	}
+
+	#[derive(Debug, Encode, Decode, TypeInfo)]
+	#[scale_info(skip_type_params(T))]
+	pub struct AssetMetadatum<T: Config> {
+		upper_bound: u16,
```

Comment:
upper_bound and lower_bound aren't currently used. but the idea is that we have limited slots (currently 20). so you can define what percentiles increase/decrease those slots would be. currently it's 1% per slot, but upper bound of like 100 would mean 5% increase per slot in the upward direction. decreases will be governed by lower bound